graylog2-web-interface/src/components/extractors/MessageLoader.tsx

```diff
@@ -0,0 +1,134 @@
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+import * as React from 'react';
+import { useCallback, useEffect, useRef, useState } from 'react';
+import PropTypes from 'prop-types';
+
+import ActionsProvider from 'injection/ActionsProvider';
+
+import { Button } from '../graylog';
+
+const MessagesActions = ActionsProvider.getActions('Messages');
+
+type LoadMessageFormProps = {
+  loadMessage: (e: React.FormEvent) => void,
+  children: React.ReactNode,
+  loading: boolean,
+};
+const LoadMessageForm = ({ loadMessage, children, loading }: LoadMessageFormProps) => (
+  <div>
+    <form className="form-inline message-loader-form" onSubmit={loadMessage}>
+      {children}
+      <Button bsStyle="info"
+              disabled={loading}
+              type="submit">
+        {loading ? 'Loading message...' : 'Load message'}
+      </Button>
+    </form>
+  </div>
+);
+
+type Props = {
+  hidden: boolean,
+  hideText: string,
+  onMessageLoaded: (data: any) => void,
+  messageId: undefined | string,
+  index: undefined | string,
+};
+
+const useMessageLoader = (defaultMessageId: string, defaultIndex: string, onMessageLoaded: (data: any) => void): [boolean, (messageId: string, index: string) => void] => {
+  const [loading, setLoading] = useState(false);
+  const loadMessage = useCallback((messageId: string, index: string) => {
+    if (messageId === '' || index === '') {
+      return;
+    }
+
+    setLoading(true);
+    const promise = MessagesActions.loadMessage(index, messageId);
+
+    promise.then((data) => onMessageLoaded(data));
+    promise.finally(() => setLoading(false));
+  }, [onMessageLoaded]);
+
+  useEffect(() => {
+    if (defaultMessageId && defaultIndex) {
+      loadMessage(defaultMessageId, defaultIndex);
+    }
+  }, [defaultMessageId, defaultIndex, loadMessage]);
+
+  return [loading, loadMessage];
+};
+
+const MessageLoader = ({ hidden, hideText, onMessageLoaded, messageId: defaultMessageId, index: defaultIndex }: Props) => {
+  const [isHidden, setIsHidden] = useState(hidden);
+
+  const [messageId, setMessageId] = useState<string>(defaultMessageId);
+  const onChangeMessageId = useCallback((e) => setMessageId(e.target.value), []);
+  const messageIdRef = useRef<HTMLInputElement>();
+
+  const [index, setIndex] = useState<string>(defaultIndex);
+  const onChangeIndex = useCallback((e) => setIndex(e.target.value), []);
+
+  const [loading, loadMessage] = useMessageLoader(defaultMessageId, defaultIndex, onMessageLoaded);
+
+  const _loadMessage = useCallback((e?: React.FormEvent) => {
+    if (e?.preventDefault) {
+      e.preventDefault();
+    }
+
+    loadMessage(messageId, index);
+  }, [index, loadMessage, messageId]);
+
+  const toggleMessageForm = useCallback(() => {
+    const newIsHidden = !isHidden;
+    setIsHidden(newIsHidden);
+
+    if (!newIsHidden) {
+      messageIdRef.current?.focus?.();
```

Comment:
This unfortunately does not work as expected. When this code runs, the `input` component is not rendered and its `ref` is therefore not set.

Before, we focussed the input after the state was set and the component was updated:

```
    this.setState({ hidden: !this.state.hidden }, this._focusMessageLoaderForm);
```

I think the best approach would be to move this if statement into its own `useEffect`, where it can run when `isHidden` changes.

Reply:
:white_check_mark: